# HW2_solutions_rokade

### Instructions
1) Start the first cell with your last name, HW/Lab number, etc. For example, HW2_solutions_basnet.

2) After you finish the HW, click File on the top left and click "download as." You will see many options. Please save the file as .ipynb and .html file.

3) To submit HW solutions, go to our course in Canvas --> Assignments --> HW 2 --> Submit Assignment --> Click "Chose file" --> choose HTML file you saved --> Click another file--> choose ipynb file--submit. Once you submit, download  and check to make sure that you have submitted the correct files.
##### ( Both HTML and ipynb files are required for your HW to be graded from now on)

4) Your submission will be compared against other student's solutions and online databases to check plagiarism.

5) You can use this ipynb file to start your solution

6) I will make HWs due on Tuesdays so that you do not have 2 assignments due on Fridays. Each problem is worth 10 points.

In [65]:
import pandas as pd
import numpy as np
columns2use = ['author_average_rating', 'author_gender','author_name','birthplace','book_average_rating',
              'book_title', 'genre_1','num_ratings', 'num_reviews', 'pages','score']

### Problem 1
Upload the grbooks.csv data in your Jupyter Notebook.  Then, select only the variables listed in the list columns2use in the previous cell and create a new data frame. Find the data type of each of the variables. Do all the variables have correct labels as objects (Categorical) or numbers? Which variable has a wrong datatype?

In [146]:
df=pd.read_csv('grbooks.csv')
df.head(8192)

author_average_rating author_gender  \
0                      4.01        female   
1                      4.15          male   
2                      4.00        female   
3                      3.88          male   
4                      4.10        female   
...                     ...           ...   
8187                   3.91          male   
8188                   3.86          male   
8189                   4.15          male   
8190                   3.98          male   
8191                   4.02          male   

                                     author_genres  author_id  \
0                              historical-fiction,      74489   
1            literature-fiction,mystery-thrillers,     706255   
2                                         romance,    5618190   
3                                  fiction,memoir,      37871   
4                             young-adult,fantasy,      36122   
...                                            ...        ...   
8187                                      romance,   13536087   
8188                      science-fiction-fantasy,       8835   
8189  science-fiction-fantasy,fiction,speculative,     130698   
8190                                  young-adult,    6449464   
8191                                       horror,     114504   

                author_name                           author_page_url  \
0       Victoria Thompson\n      /author/show/74489.Victoria_Thompson   
1           Stieg Larsson\n         /author/show/706255.Stieg_Larsson   
2     Mimi Jean Pamfiloff\n  /author/show/5618190.Mimi_Jean_Pamfiloff   
3             José Donoso\n             /author/show/37871.Jos_Donoso   
4       Patricia C. Wrede\n       /author/show/36122.Patricia_C_Wrede   
...                     ...                                       ...   
8187        Paige Laurens\n       /author/show/13536087.Paige_Laurens   
8188           Tim Powers\n              /author/show/8835.Tim_Powers   
8189           Ted Chiang\n            /author/show/130698.Ted_Chiang   
8190       Tracie Puckett\n       /author/show/6449464.Tracie_Puckett   
8191         Scott Sigler\n          /author/show/114504.Scott_Sigler   

      author_rating_count  author_review_count                birthplace  \
0                   74399                 6268       United States\n       
1                 3726435               142704              Sweden\n       
2                   76496                 7975         United States\n     
3                    5522                  489               Chile\n       
4                  291013                13453       United States\n       
...                   ...                  ...                       ...   
8187                 1408                  139         United States\n     
8188                63860                 6052       United States\n       
8189                62069                 7887         United States\n     
8190                24166                 1305    \n   United States\n     
8191                83990                 8268  \n   United States\n       

      book_average_rating                                       book_fullurl  \
0                    4.02  https://www.goodreads.com/book/show/686717.Mur...   
1                    4.13  https://www.goodreads.com/book/show/2429135.Th...   
2                    3.99  https://www.goodreads.com/book/show/27833684-t...   
3                    4.14  https://www.goodreads.com/book/show/382975.The...   
4                    4.01  https://www.goodreads.com/book/show/64207.Sorc...   
...                   ...                                                ...   
8187                 3.96  https://www.goodreads.com/book/show/25004287-b...   
8188                 3.94  https://www.goodreads.com/book/show/142296.The...   
8189                 4.26  https://www.goodreads.com/book/show/223380.Sto...   
8190                 4.23  https://www.goodreads.com/book/show/17876011-t...   
8191          

In [136]:
df1=df[columns2use]
df1.head(4)

author_average_rating author_gender            author_name  \
0                   4.01        female    Victoria Thompson\n   
1                   4.15          male        Stieg Larsson\n   
2                   4.00        female  Mimi Jean Pamfiloff\n   
3                   3.88          male          José Donoso\n   

             birthplace  book_average_rating  \
0   United States\n                     4.02   
1          Sweden\n                     4.13   
2     United States\n                   3.99   
3           Chile\n                     4.14   

                                  book_title  genre_1  num_ratings  \
0       \n      Murder on St. Mark's Place\n  Mystery         5260   
1  \n      The Girl with the Dragon Tattoo\n  Fiction      2229163   
2             \n      Tailored for Trouble\n  Romance         2151   
3        \n      The Obscene Bird of Night\n  Fiction         1844   

   num_reviews pages  score  
0          375   277   3230  
1        65227   465   3062  
2          391   354   4585  
3          173   438   1533

In [137]:
df1.dtypes

author_average_rating    float64
author_gender             object
author_name               object
birthplace                object
book_average_rating      float64
book_title                object
genre_1                   object
num_ratings                int64
num_reviews                int64
pages                     object
score                      int64
dtype: object

Pages is haing a datatype object when it should have int as datatype.

### Problem 2
Continuing with the data you obtained in problem 1, find the $\textbf{median}$ of all the numerical variables you saw in problem 1. Write a paragraph commenting on the median of these variables. 

In [68]:
median=df1.median(numeric_only=True)
median

author_average_rating       3.97
book_average_rating         3.96
num_ratings              4403.00
num_reviews               384.00
score                    1727.00
dtype: float64

The median values show that the author rating of 3.97 correlates to a book rating of 3.96, i.e, if the author is good, then the book is highly rated. The number of median ratings is 4403 whereas the number of median reviews is a lot less, i.e., 384. The median score is 1727

### Problem 3 
Find the $\textbf{mode}$ of all the numerical variables you saw in problem 1. Write a paragraph commenting on the mode of these variables. 

In [69]:
mode=df1.mode(numeric_only=True)
mode

author_average_rating  book_average_rating  num_ratings  num_reviews  score
0                   3.94                 3.96           44           17    200

The mode values show that the author rating of 3.94 has the max. frequency throughout the column. The book rating of 3.96 has max. frequency in this dataframe.The max. frequency of ratings is 44, while the max. frequency of number of ratings and score is 17 and 200 resepctively.

### Problem 4
A distribution (here it is one of the numerical columns) in which the values of mean, median, and mode coincide (i.e., mean = median = mode) is known as a normal distribution (a perfect data set).It has a bell shapped curve when plotted and is also symmetric about the mean. Conversely, when mean, median, and mode values are not equal, the distribution is known as asymmetrical or skewed distribution. In moderately skewed or asymmetrical distribution, a very important relationship exists among these three measures of central tendencies. In such distributions, the distance between the mean and median is about one-third of the mean and mode distance. Karl Pearson, a renowned statistician, expressed this relationship as:
$Mode - 3 \times \,\,median - 2\times \,\, mean =0$. 

![](mcube.png) 


Find Mode - 3 Median + 2 Mean for all numerical columns. Symmetric data is hard to come by. If this number is close to 0 for a column, the column data is close to being symmetric, and if it is far from 0 ( positive or negative), the column data is asymmetric. Which column is closest to being symmetric, and which one is the most asymmetric?

In [70]:
mean=df1.mean(numeric_only=True)
mean

author_average_rating        3.960368
book_average_rating          3.951456
num_ratings              46683.487790
num_reviews               2324.754794
score                     3893.028483
dtype: float64

In [71]:
formula=(mode-3)*(median-2)*(mean)

formula

author_average_rating  book_average_rating   num_ratings   num_reviews  \
0                7.33381              7.43506  8.423615e+09  1.243279e+07   

          score  
0  1.322948e+09

In [72]:
formula['score'] = formula['score'].astype('int64')
formula['num_ratings'] = formula['num_ratings'].astype('int64')
formula['num_reviews'] = formula['num_reviews'].astype('int64')
formula

author_average_rating  book_average_rating  num_ratings  num_reviews  \
0                7.33381              7.43506   8423615220     12432788   

        score  
0  1322948404

author_average_rating is the most symmetrical variable in this dataframe, while score is the most assymetric

### Problem 5
We can use sample function from pandas to randomly select only few  rows from our data frame. Sampling is very useful when the data is too big and not all of it can be handled in a computer.
More about sample function is explained here.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html


Using the sample function, sample 1000 data points from the data you created in problem 1 and give it a new name. Now find the means  of this new sampled data and compare with the means of the data you sampled from. Which variable's mean  from this sampled data setis close to the mean of the original data? What does it imply when these two means are very close? 

In [73]:
sampledf=df1.sample(n=1000).reset_index()
sampledf.head(4)

index  author_average_rating author_gender                 author_name  \
0   6049                   4.04        female         Alyssa Rae Taylor\n   
1    402                   4.26        female           Patricia Briggs\n   
2  17731                   3.85          male          Andrew C. Walton\n   
3  16013                   4.29          male  Douglas Southall Freeman\n   

                 birthplace  book_average_rating  \
0  \n   United States\n                     4.01   
1         United States\n                   4.18   
2                      \n                   3.81   
3       United States\n                     4.42   

                                          book_title       genre_1  \
0                            \n      Raising Ryann\n       Romance   
1                              \n      Moon Called\n       Fantasy   
2  \n      The Spiritual Secrets Of Happiness Hea...  Spirituality   
3    \n      Lee's Lieutenants: A Study In Command\n       History   

   num_ratings  num_reviews pages  score  
0         9695          600   277   6413  
1       151855         7321   288  17039  
2          220            6   196    505  
3          358            9   832    606

In [74]:
sampledf.mean(numeric_only=True)

index                    11420.52400
author_average_rating        3.96518
book_average_rating          3.95597
num_ratings              46022.59100
num_reviews               2523.04900
score                     3915.02900
dtype: float64

In [75]:
mean

author_average_rating        3.960368
book_average_rating          3.951456
num_ratings              46683.487790
num_reviews               2324.754794
score                     3893.028483
dtype: float64

The book_average_rating and author_average_rating have almost identical mean values. This shows that the randomly sampled data has approximately the same values from the main dataframe. This reinforces our previous idea that these variables are almost symmetric and most symmetric within this dataframe.

### Problem 6
Make a frequency table of the variable "genre_1". Your answer should look like a data frame with indices ( with no column name), and two other column names "genre_1" and "frequency". The following Panda functions will be useful: groubby, value_counts, reset_index() and to_frame(). https://stackoverflow.com/questions/39778686/pandas-reset-index-after-groupby-value-counts and https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html

In [110]:
freq_table=df1['genre_1'].value_counts().to_frame().reset_index()
freq_table.columns = ['genre_1','counts']
freq_table

genre_1  counts
0         Fantasy    3397
1         Romance    3397
2     Young Adult    2497
3         Fiction    2172
4      Historical    1399
..            ...     ...
119       Medical       1
120         Death       1
121  Dark Fantasy       1
122      Teaching       1
123     Dc Comics       1

[124 rows x 2 columns]

### Problem 7. 
Using the column  "score", to create a new column called new_score that will have a score out of 5. For this, apply the formula  $\text{new_score} = \dfrac{\text{score}}{\text{max(score)}}\times 5$. Print the description of this new colum using describe function.

In [115]:
df1["new_score"]=df1.score/max(df1.score)*5

<ipython-input-115-9e99b9e232d4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1["new_score"]=df1.score/max(df1.score)*5


In [112]:
df1.new_score.describe()

count    22891.000000
mean         0.032536
std          0.092116
min          0.000460
25%          0.006953
50%          0.014433
75%          0.030066
max          5.000000
Name: new_score, dtype: float64

### Problem 8.
We have now 3 columns that represent ratings out of 5, namely author_average_rating, book_average_rating, and a new_score we just created in problem 7. Let's create a column called super_score which takes into consideration of all the 3 columns with rating scores out of 5. Give 24%, 34% and 42% weights for the columns author_average_rating, book_average_rating, and a new_score respectively to create the values for super_score column that is also out of 5. Now use the describe function to show the details of this new column.

In [116]:
df1['super_score']=(0.24*df1.author_average_rating)+(0.34*df1.book_average_rating)+(0.42*df1.new_score)

<ipython-input-116-19896c8e67db>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['super_score']=(0.24*df1.author_average_rating)+(0.34*df1.book_average_rating)+(0.42*df1.new_score)


In [117]:
df1.super_score.describe()

count    22891.000000
mean         2.307648
std          0.156127
min          0.871881
25%          2.211681
50%          2.310253
75%          2.406540
max          4.625000
Name: super_score, dtype: float64

### Problem 9
Use super_score column to find out how many unique authors have a super_score of 3.9 or more. Let's call these authors, super_authors. Create a dataframe called super_authors that has the following columns: ['author_name','author_gender','birthplace','book_title'] and show the head of the data.

In [122]:
df1["super_authors"]=df1.super_score>=3.9

<ipython-input-122-7d1c4f4019d6>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1["super_authors"]=df1.super_score>=3.9


In [124]:
df1.super_authors.value_counts()

False    22888
True         3
Name: super_authors, dtype: int64

In [125]:
super_authors = df1[['author_name','author_gender','birthplace','book_title']].copy()

In [126]:
super_authors

author_name author_gender              birthplace  \
0        Victoria Thompson\n        female     United States\n       
1            Stieg Larsson\n          male            Sweden\n       
2      Mimi Jean Pamfiloff\n        female       United States\n     
3              José Donoso\n          male             Chile\n       
4        Patricia C. Wrede\n        female     United States\n       
...                      ...           ...                     ...   
22886       Dakota Cassidy\n        female                    \n     
22887      Joanne Jaytanie\n          male     United States\n       
22888      Krystal Shannan\n        female       United States\n     
22889          Bryce Evans\n        female                    \n     
22890     Candice Stauffer\n          male  \n   United States\n     

                                              book_title  
0                   \n      Murder on St. Mark's Place\n  
1              \n      The Girl with the Dragon Tattoo\n  
2                         \n      Tailored for Trouble\n  
3                    \n      The Obscene Bird of Night\n  
4      \n      Sorcery & Cecelia: or The Enchanted Ch...  
...                                                  ...  
22886                         \n      Gotta Have Faith\n  
22887                         \n      Payton's Pursuit\n  
22888                        \n      My Viking Vampire\n  
22889                              \n      The Trinity\n  
22890              \n      Volatile Breath of Darkness\n  

[22891 rows x 4 columns]

Therefore, 3 authors have a super score above 3.9

### Problem 10
Save the dataframe super_authors you created in problem 9 to your working folder (where you have this HW saved) as a csv file and read it back to Jupyter Notebook and print the head of the data.

In [129]:
super_authors.to_csv('super_authors.csv')

In [130]:
df2=pd.read_csv('super_authors.csv')
df2.head()

Unnamed: 0            author_name author_gender            birthplace  \
0           0    Victoria Thompson\n        female   United States\n       
1           1        Stieg Larsson\n          male          Sweden\n       
2           2  Mimi Jean Pamfiloff\n        female     United States\n     
3           3          José Donoso\n          male           Chile\n       
4           4    Patricia C. Wrede\n        female   United States\n       

                                          book_title  
0               \n      Murder on St. Mark's Place\n  
1          \n      The Girl with the Dragon Tattoo\n  
2                     \n      Tailored for Trouble\n  
3                \n      The Obscene Bird of Night\n  
4  \n      Sorcery & Cecelia: or The Enchanted Ch...

## The following 2 problems are extra credit problems and are optional. You won't get any help on these problems.

### Bonus Problem 1 (Bonus 10 points)
In super_author data set, there are a few columns that have entries with $\text{\n}$ notation. Find these columns and remove $\text{\n}$ from all the column entries using regular expressions.  Read more about handling strings and regular expressions here. https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html, https://www.geeksforgeeks.org/replace-values-in-pandas-dataframe-using-regex/
, and here https://docs.python.org/2/library/re.html.

In [141]:
super_authors = super_authors.replace(r'\n',' ', regex=True)
super_authors

author_name author_gender            birthplace  \
0        Victoria Thompson         female    United States        
1            Stieg Larsson           male           Sweden        
2      Mimi Jean Pamfiloff         female      United States      
3              José Donoso           male            Chile        
4        Patricia C. Wrede         female    United States        
...                     ...           ...                   ...   
22886       Dakota Cassidy         female                         
22887      Joanne Jaytanie           male    United States        
22888      Krystal Shannan         female      United States      
22889          Bryce Evans         female                         
22890     Candice Stauffer           male      United States      

                                              book_title  
0                            Murder on St. Mark's Place   
1                       The Girl with the Dragon Tattoo   
2                                  Tailored for Trouble   
3                             The Obscene Bird of Night   
4             Sorcery & Cecelia: or The Enchanted Cho...  
...                                                  ...  
22886                                  Gotta Have Faith   
22887                                  Payton's Pursuit   
22888                                 My Viking Vampire   
22889                                       The Trinity   
22890                       Volatile Breath of Darkness   

[22891 rows x 4 columns]

### Bonus Problem 2 (Bonus 10 points)
You may have noticed that the column named  "pages" from the data set in problem 1 seem to have numerical looking values but it is not showing up in our data summary. This is because this columns appears to be a string in our data set. Data values in this columns are just  the total number of pages in each book and these should be numbers not strings. Change this column from string to numeric and find mean, median and mode for this column. If the error comes in the process of converting to numerical, find out what the error is and fix it.

In [144]:
pd.to_numeric(df['pages'])

ValueError: Unable to parse string "1 page" at position 8191

In [147]:
df['pages'] = df['pages'].replace(['1 page'],'1')

In [148]:
pd.to_numeric(df['pages'])

0        277
1        465
2        354
3        438
4        326
        ... 
22886     72
22887    342
22888    237
22889    222
22890    272
Name: pages, Length: 22891, dtype: int64

In [159]:
df.pages.mean()

inf

In [160]:
df.pages.median()

315.0

In [161]:
df.pages.mode()

0    320
dtype: object